### PR TITLE
refactor: replace glu.Config with glu.PipelineBuilder

### PR DESCRIPTION
**Pull request by GeorgeMac**
Continuing to seize on the pre release and breakable nature of this project 😂 

This removes the concept of `glu.Config` and replaces it with an optional `glu.PipelineBuilder`.
The builder is where the (now optional) typed resource comes into play.

Previously, we couldn't hang a type parameter off the core system, otherwise, we tie the system to a single type parameter.
Same went for configuration and the pipeline interface at this fidelity.

This made creating a config structure that could simplify configuring a typed Git and OCI source hard to do.
Now, `AddPipeline` takes a `func(context.Context, *config.Config) (Pipeline, error)`.
This is not typed and only has access to the raw config.

The new function `glu.BuilderFunc` returns this function type, but it does so by adapting a function which takes the new `*PiplineBuilder` type.

The two functions previously hung off `*glu.Config`, called `GitRepository` and `OCIRepository` have now morphed into:
```
glu.GitSource(...)
// and
glu.OCISource(...)
```

These two functions depend on this builder (and they continue to use it for caching previouslly built dependencies) and return the actual typed `git.Source[R]` and `oci.Source[R]`.
Each of these functions and types share the core constraint of `glu.Resource`.
However, the Git and OCI ones further constrain it with their respective necessary requirements.

TL;DR this removes the need to know about a sources repository dependency (and removes the need for more type arguments):

https://github.com/get-glu/glu/blob/e3daa16e6d2b57608f8badc0bef1f75550d5be4f/examples/oci-and-git/experiment.go#L22-L37

### CODE REVIEW

**Comment by GeorgeMac** [comment]
Hold this thought. I think this might bork caching.

**Comment by GeorgeMac** [comment]
I think there is something here, but it does break the cache (caching is no longer on the single system).
I am going to take another pass at this later.
I think there is something here, just the result will be a little more invasive.
